crypto/s2n_rsa_pss.c

```diff
@@ -0,0 +1,216 @@
+}
+
+/* On some versions of OpenSSL, "EVP_PKEY_CTX_set_signature_md()" is just a macro that casts digest_alg to "void*",
+ * which fails to compile when the "-Werror=cast-qual" compiler flag is enabled. So we work around this OpenSSL
+ * issue by creating a local non-const duplicate of EVP_MD pointer, calling OpenSSL with the duplicate, then freeing the
+ * duplicate since our compiler won't let us pass a const pointer to an API that doesn't guarantee to not modify it. */
+static int s2n_evp_ctx_set_signature_digest(EVP_PKEY_CTX *ctx, const EVP_MD* const_digest_alg, int s2n_err) {
+    notnull_check(ctx);
+    notnull_check(const_digest_alg);
+
+    EVP_MD* digest_alg = EVP_MD_meth_dup(const_digest_alg);
+    notnull_check(digest_alg);
+
+    int rc = EVP_PKEY_CTX_set_signature_md(ctx, digest_alg);
+    EVP_MD_meth_free(digest_alg);
+    S2N_ERROR_IF(rc <= 0, s2n_err);
+
+    return 0;
+}
+
+static int s2n_rsa_pss_sign(const struct s2n_pkey *priv, struct s2n_hash_state *digest, struct s2n_blob *signature_out)
+{
+    notnull_check(priv);
+
+    uint8_t digest_length;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_digest_size(digest->alg, &digest_length));
+    GUARD(s2n_hash_digest(digest, digest_data, digest_length));
+
+    const EVP_MD* digest_alg = s2n_hash_alg_to_evp_alg(digest->alg);
+    notnull_check(digest_alg);
+
+    /* For more info see: https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_sign.html */
+    EVP_PKEY_CTX *ctx  = EVP_PKEY_CTX_new(priv->key.rsa_pss_key.pkey, NULL);
+    notnull_check(ctx);
+
+    size_t signature_len = signature_out->size;
+    GUARD_OSSL(EVP_PKEY_sign_init(ctx), S2N_ERR_SIGN);
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING), S2N_ERR_SIGN);
+    GUARD(s2n_evp_ctx_set_signature_digest(ctx, digest_alg, S2N_ERR_SIGN));
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_pss_saltlen(ctx, RSA_PSS_SALTLEN_DIGEST), S2N_ERR_SIGN);
+
+    /* Calling EVP_PKEY_sign() with NULL will only update the signature_len parameter so users can validate sizes. */
+    GUARD_OSSL(EVP_PKEY_sign(ctx, NULL, &signature_len, digest_data, digest_length), S2N_ERR_SIGN);
+    S2N_ERROR_IF(signature_len > signature_out->size, S2N_ERR_SIZE_MISMATCH);
+
+    /* Actually sign the the digest */
+    GUARD_OSSL(EVP_PKEY_sign(ctx, signature_out->data, &signature_len, digest_data, digest_length), S2N_ERR_SIGN);
+    signature_out->size = signature_len;
+
+    return 0;
+}
+
+static int s2n_rsa_pss_verify(const struct s2n_pkey *pub, struct s2n_hash_state *digest, struct s2n_blob *signature_in)
+{
+    notnull_check(pub);
+
+    uint8_t digest_length;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_digest_size(digest->alg, &digest_length));
+    GUARD(s2n_hash_digest(digest, digest_data, digest_length));
+    const EVP_MD* digest_alg = s2n_hash_alg_to_evp_alg(digest->alg);
+
+    /* For more info see: https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_verify.html */
```

Comment:
This comment is suited for line 133.

Reply:
I intended this comment to cover lines 127-133 since those lines are similar to OpenSSL's sample code in the link.